equations/networking/game_flow.py

```diff
@@ -328,3 +353,39 @@ def handle_bonus_click():
 def handle_call_judge():
     """Player requested to call a judge."""
     pass
+
+@equations.socketio.on("variation_called")
+def handle_variation_called(info):
+    """A player submitted variations."""
+    MapsLock()
+    [name, room] = get_name_and_room(flask.request.sid)
+
+    assert info["player"] == name
+
+    # Split provided string into variations. A variation in the string is considered
+    # to be a consecutive sequence of numbers, letters, underscore, hyphen, and apostrophe.
+    # This operation also serves to sufficiently sanitize what's being added to the DOM.
+    variations = re.findall(r"[-\w']+", info["content"])
```

Comment:
regex is only for letters, numbers, hyphen, and apostrophes (underscore is missing)

other issues:
- can't just enter "!" for factorial (inserts blank and considers you "done" with calling variations)
    - check input to see if the string that the player types maps to empty string and prompt them to input something that maps to a string
    - maybe use comma as a separator? right now it is using space or anything that isn't [-\w'] as a separator and using comma would be most natural for most players
- typing "0 wild" results in "0, wild" because of the last bullet point above ^
- in a 2 player game, players should probably be prompted by the server for who is supposed to call 2 variations

Reply:
Good points. 

One question: Why do you think we should match on underscores?

Reply:
Also just to confirm each player has to call at least one variation right?

Reply:
I fixed some of these bugs directly onto master. Take a look and let me know what you think: https://github.com/tonyb7/equations/blob/master/equations/networking/game_flow.py#L365-L379

I didn't bother to add better server prompts yet (like how many variations to call), since if we do that, then we would also have to implement dividing the room based on division, since in Senior players can call up to 4 variations iirc. So right now the only restriction is that a player has to call at least one variation, and the prompt just tells players to "enter their variations" (which is generic and could refer to however many they are supposed to enter).

Feel free to implement more features for variation-calling if you have ideas and the time though.

Reply:
You wrote underscore in the comment on line 366, which doesn't match the previous or current regex on line 368 (comment isn't in sync with the code that it's referring to)
Making each player call at least 1 variation should be good enough. We'll probably just have to deal with people forgetting to call 2 variations in 2 player games.

Reply:
I see. Fixed https://github.com/tonyb7/equations/commit/98cee9297850e017f3de9b8f24156f3ebae3a12d